openedx/core/djangoapps/lang_pref/middleware.py

```diff
@@ -27,9 +29,12 @@ def process_request(self, request):
         If a user's UserPreference contains a language preference, use the user's preference.
         Save the current language preference cookie as the user's preferred language.
         """
-        cookie_lang = request.COOKIES.get(settings.LANGUAGE_COOKIE, None)
+        cookie_lang = request.COOKIES.get(settings.LANGUAGE_COOKIE_NAME, None)
         if cookie_lang:
             if request.user.is_authenticated:
+                # DarkLangMiddleware has already taken care of this so don't change anything
+                if DarkLangConfig.current().enabled and get_user_preference(request.user, DARK_LANGUAGE_KEY):
+                    return
                 set_user_preference(request.user, LANGUAGE_KEY, cookie_lang)
```

Comment:
Am I understanding the existing code correctly? *Every* authenticated browser request to the LMS results in a DB call to update the user's language preferences based on the existing cookie, and then gets back a new version of the same cookie? This seems like a very strange way of going about things, and I wonder why it's like that.

This must cause some strange bugs. For example, I was able to produce a "contagious language" behavior:

1. Log into a first edX account and change language to Korean
2. Log out
3. Without clearing cookies, log into a second edX account that was previously English -- it now shows the interface in Korean, because `prod-edx-language-preference` didn't get cleared during logout of the previous session.
4. On a *different* computer, with no cookies present, log into that second account again. It shows Korean now here as well!

Regardless of what we do for this Django 3.0 PR, I feel like this whole language preference thing needs to be replaced by something less contorted.

Reply:
I was able to reproduce this behaviour on stage as well. As we are nearing Django 3.2 upgrade deadline, I think we can merge this for now and we can create a ticket to fix this. Also, we are using both the `Accept-Language` header and the cookie approach for setting language, we can remove one of these to reduce the complexity. So it definitely needs further work for fixing language-related bugs.

Reply:
Yeah, I think later we'll want to collapse these two middlewares together and use a simpler approach, but we need to find out more about the requirements before that.